### PR TITLE
CANDLEPIN-718: Re-design the certificate generation

### PR DESCRIPTION
**Pull request by Januson**

### CODE REVIEW

File: src/main/java/org/candlepin/pki/huffman/PathNode.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        }
+        // same number of children with the same names for child nodes
+        if (this.getChildren().size() != that.getChildren().size()) {
+            return false;
+        }
+        for (NodePair thisnp : this.getChildren()) {
+            boolean found = false;
+            for (NodePair thatnp : that.getChildren()) {
+                if (thisnp.getName().equals(thatnp.getName())) {
+                    if (thisnp.getConnection().isEquivalentTo(thatnp.getConnection())) {
```
Comment (joshmalbrecht) [inline src/main/java/org/candlepin/pki/huffman/PathNode.java:76]: You can && these two conditions into one if statement if you want to.  I think this is more of preference thing and so keeping as is also works.

File: src/main/java/org/candlepin/pki/huffman/NodePair.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    public PathNode getConnection() {
+        return connection;
+    }
+
+    void setConnection(PathNode connection) {
+        this.connection = connection;
+    }
+
+    public String toString() {
```
Comment (joshmalbrecht) [inline src/main/java/org/candlepin/pki/huffman/NodePair.java:39]: Should the `toString`, `equals`, and `hashcode` methods have the override annotation?

File: src/main/java/org/candlepin/pki/huffman/Huffman.java [hunk 3]
```diff
@@ -0,0 +1,677 @@
...
+
+public class Huffman {
+    private static final Logger log = LoggerFactory.getLogger(Huffman.class);
+    private static final Object END_NODE = new Object();
+    private static final boolean TREE_DEBUG = false;
+
+    private long pathNodeId = 0;
+    private long huffNodeId = 0;
+
+    @Inject
```
Comment (joshmalbrecht) [inline src/main/java/org/candlepin/pki/huffman/Huffman.java:50]: I think we can remove this inject annotation if there are no objects being injected.

File: src/main/java/org/candlepin/pki/huffman/PathNode.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                }
+            }
+            if (!found) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public String toString() {
```
Comment (joshmalbrecht) [inline src/main/java/org/candlepin/pki/huffman/PathNode.java:92]: Does this need override annotation?

File: src/main/java/org/candlepin/pki/certs/X509CertificateBuilder.java [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+        this.addKeyUsage(builder);
+        this.addAuthorityKeyIdentifier(builder, caCertificate);
+        this.addSubjectKeyIdentifier(builder, this.keyPair);
+        this.addSubjectAltName(builder, this.distinguishedName, this.subjectAltName);
+        this.addBasicConstraints(builder);
+        this.addExtensions(builder, this.certExtensions);
+
+        return buildCertificate(builder, this.signer());
+    }
+
```
Comment (joshmalbrecht) [inline src/main/java/org/candlepin/pki/certs/X509CertificateBuilder.java:177]: Extra new line

**Review by joshmalbrecht — CHANGES_REQUESTED** [review]
Just noticed some small stuff and had a couple questions

File: src/main/java/org/candlepin/pki/certs/AnonymousCertificateGenerator.java [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+                converted.setRequiredTags(content.getRequiredTags());
+                converted.setArches(content.getArches());
+                converted.setContentUrl(content.getContentUrl());
+                converted.setReleaseVersion(content.getReleaseVersion());
+
+                return converted;
+            })
+            .toList();
+    }
+
```
[the remark this answers is not in the document]
Comment (Januson) [reply src/main/java/org/candlepin/pki/certs/AnonymousCertificateGenerator.java:342]: Done.

File: src/main/java/org/candlepin/async/tasks/RegenProductEntitlementCertsJob.java [hunk 7]
```diff
@@ -49,7 +49,7 @@ public class RegenProductEntitlementCertsJob implements AsyncJob {
     private static final String ARG_PRODUCT_ID = "product_id";
     private static final String ARG_LAZY_REGEN = "lazy_regen";
 
-    private final EntitlementCertificateGenerator ecGenerator;
+    private final EntitlementCertificateService ecGenerator;
```
[the remark this answers is not in the document]
Comment (Januson) [reply src/main/java/org/candlepin/async/tasks/RegenProductEntitlementCertsJob.java]: Done.

File: src/main/java/org/candlepin/pki/certs/EntitlementPayloadGenerator.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        toReturn.setConsumer(consumerUuid);
+        toReturn.setQuantity(quantity);
+        toReturn.setSubscription(createSubscription(pool));
+        toReturn.setOrder(createOrder(pool));
+        toReturn.setProducts(productModels);
+        toReturn.setPool(createPool(pool));
+
+        return toReturn;
+    }
+
```
[the remark this answers is not in the document]
Comment (Januson) [reply src/main/java/org/candlepin/pki/certs/EntitlementPayloadGenerator.java:78]: Done.

File: src/main/java/org/candlepin/pki/huffman/PathNode.java (hunk 1 above)
Comment (Januson) [reply src/main/java/org/candlepin/pki/huffman/PathNode.java]: Done. I wasn't happy with readability of the combined check so I ended up extracting it into a method.

File: src/main/java/org/candlepin/pki/huffman/NodePair.java (hunk 2 above)
Comment (Januson) [reply src/main/java/org/candlepin/pki/huffman/NodePair.java]: Done.

File: src/main/java/org/candlepin/pki/huffman/Huffman.java (hunk 3 above)
Comment (Januson) [reply src/main/java/org/candlepin/pki/huffman/Huffman.java]: Done.

File: src/main/java/org/candlepin/pki/huffman/PathNode.java (hunk 4 above)
Comment (Januson) [reply src/main/java/org/candlepin/pki/huffman/PathNode.java:96]: Done.

File: src/main/java/org/candlepin/pki/certs/X509CertificateBuilder.java (hunk 5 above)
Comment (Januson) [reply src/main/java/org/candlepin/pki/certs/X509CertificateBuilder.java:176]: Done.

File: src/main/java/org/candlepin/pki/huffman/PathNode.java (hunk 1 above)
Comment (joshmalbrecht) [reply src/main/java/org/candlepin/pki/huffman/PathNode.java]: Looks good!

**Pull request closed without merging**